datumaro/plugins/wider_format.py

```diff
@@ -0,0 +1,115 @@
+    IMAGE_EXT = '.jpg'
+    ANNOTATIONS_DIR = 'wider_face_split'
+    IMAGES_DIR = 'images'
+    SUBSET_DIR = 'WIDER_'
+    BBOX_ATTRIBUTES = ['blur', 'expression', 'illumination',
+        'occlusion', 'pose', 'invalid']
+
+class WiderExtractor(SourceExtractor):
+    def __init__(self, path):
+        assert osp.isfile(path), path
+        self._path = path
+        self._dataset_dir = osp.dirname(osp.dirname(path))
+
+        subset = osp.splitext(osp.basename(path))[0]
+        match = re.fullmatch(r'wider_face_\S+_bbx_gt', subset)
+        if match:
+            subset = subset.split('_')[2]
+        super().__init__(subset=subset)
+
+        self._items = list(self._load_items(path).values())
+
+    def _load_items(self, path):
+        items = {}
+        with open(path, 'r') as f:
+            lines = f.readlines()
+
+        image_ids = [image_id for image_id, line in enumerate(lines)
+            if WiderPath.IMAGE_EXT in line]
+
+        for image_id in image_ids:
+            image = lines[image_id]
+            image_path = osp.join(self._dataset_dir, WiderPath.SUBSET_DIR
+                + self._subset, WiderPath.IMAGES_DIR, image[:-1])
+            item_id = image[:-(len(WiderPath.IMAGE_EXT) + 1)]
+
+            bbox_count = lines[image_id + 1]
+            bbox_lines = lines[image_id + 2:image_id + 2 + int(bbox_count)]
+            annotations = []
+            for bbox in bbox_lines:
+                bbox_list = bbox.split()
+                if len(bbox_list) >= 4:
+                    attributes = {}
+                    if len(bbox_list) == 10:
+                        i = 4
+                        for attr in WiderPath.BBOX_ATTRIBUTES:
+                            attributes[attr] = int(bbox_list[i])
+                            i += 1
+                    annotations.append(Bbox(
+                        int(bbox_list[0]), int(bbox_list[1]),
+                        int(bbox_list[2]), int(bbox_list[3]),
+                        attributes = attributes
+                    ))
+
+            items[item_id] = DatasetItem(id=item_id, subset=self._subset,
+                image=image_path, annotations=annotations)
+        return items
+
+class WiderImporter(Importer):
+    @classmethod
+    def find_sources(cls, path):
+        return cls._find_sources_recursive(osp.join(path,
+            WiderPath.ANNOTATIONS_DIR), '.txt', 'wider')
+
+class WiderConverter(Converter):
+    DEFAULT_IMAGE_EXT = '.jpg'
+
+    def apply(self):
+        save_dir = self._save_dir
+
+        os.makedirs(save_dir, exist_ok=True)
+
+        for subset_name, subset in self._extractor.subsets().items():
+            subset_dir = osp.join(save_dir, WiderPath.SUBSET_DIR + subset_name)
+
+            wider_annotation = ''
+            for item in subset:
+                wider_annotation += '%s\n' % (item.id + WiderPath.IMAGE_EXT)
+                if self._save_images:
+                    self._save_image(item, osp.join(save_dir, subset_dir,
```

Comment:
Please, add `.image` existence check with `.has_image`.

Reply:
Why add this check if it already exists in `_save_image`?

Reply:
`_save_image` checks for image data, but it expects an image. It is possible that input data does not have image info at all, just annotations. A good question, though. Maybe, this check is worth adding there as well.